cmd/login.go

```diff
@@ -102,208 +112,157 @@ type deviceTokenResponse struct {
-					"client_id":     {authProvider.ClientID},
-					"refresh_token": {ctxt.RefreshToken}})
-
-			if err != nil {
-				return "", fmt.Errorf("Cannot refresh access token - %s", err)
+			params := url.Values{
+				"refresh_token": {ctxt.RefreshToken},
 			}
-
-			var tokenResponse deviceTokenResponse
-			jsonDecoder := json.NewDecoder(response.Body)
-			if err := jsonDecoder.Decode(&tokenResponse); err != nil {
-				return "", fmt.Errorf("Cannot decode token response - %s", err)
+			tokenResponse := getTokenResponse(authProvider, params, ctxt, false)
+			if tokenResponse.ErrorString != "" {
+				logger.Warn("tokenResponse", log.String("error", tokenResponse.ErrorString))
+				cobra.CheckErr("oauth: Unexpected error from authentication provider")
 			}
 
-			switch tokenResponse.ErrorString {
-			case "authorization_pending":
-				// No op - we're waiting on the user to open the link and login
-			case "expired_token":
-				return "", fmt.Errorf("The login process was not completed in time - please login again")
-			case "access_denied":
-				return "", fmt.Errorf("Could not login - access was denied")
-			case "invalid_grant":
-				return "", fmt.Errorf("Could not login - expired credentials. Please use the login command to refresh your credentials")
-			case "":
-				// No Errors:
-				ctxt.AccessToken = tokenResponse.AccessToken
-				// Add a 10 second buffer to expiry to account for differences in clock time between client
-				// server and message transport time (oauth2 library does the same thing)
-				ctxt.AccessTokenExpiry = time.Now().Add(time.Second * time.Duration(tokenResponse.ExpiresIn-10))
-
-				// We also get an updated ID token, let's make sure we have the latest info
-				ParseIDToken(&tokenResponse, ctxt, authProvider.JwksURL)
-
-				fmt.Println(fmt.Sprintf("Successfully acquired new access token. Expiry: %s", ctxt.AccessTokenExpiry))
-
-				SetContext(ctxt, true)
-			}
+			ctxt.AccessToken = tokenResponse.AccessToken
+			ctxt.RefreshToken = tokenResponse.RefreshToken
+			// Add a 10 second buffer to expiry to account for differences in clock time between client
+			// server and message transport time (oauth2 library does the same thing)
+			ctxt.AccessTokenExpiry = time.Now().Add(time.Second * time.Duration(tokenResponse.ExpiresIn-10))
 
+			// We also get an updated ID token, let's make sure we have the latest info
+			ParseIDToken(&tokenResponse, ctxt, authProvider.JwksURL)
+			SetContext(ctxt, true)
+			fmt.Printf("Successfully acquired new access token. Expiry: %s", ctxt.AccessTokenExpiry.Format(time.RFC822))
 		} // Access token has not expired, let's just use it
 	}
 
-	return ctxt.AccessToken, nil
+	return ctxt.AccessToken
 }
 
-func getLoginInformation(client *http.Client, ctxt *Context) (authProvider *AuthProvider, err error) {
-	if ctxt == nil {
-		return nil, fmt.Errorf("Invalid config set. Please set a valid config with the config command.")
+func IsAuthorised() bool {
```

Comment:
This logic does not match the logic in the adapter (which also checks for the environmental variable). We may also have an access token but it is out of date, and therefore we're no longer authorised.

Reply:
OK, that's an interesting on. If we pick up the access token via an environment variable we would need to parse it and check the valid-to field. 

This part of the code is used for information purposes only (context.Get), so how accurate do we want to be?

If we pick it up via an ENV varibale do we just print - "don't really know?"

Reply:
My expectation for the use of the ENV variable is in scripts where we pass it in via some other mechanism.

Reply:
If it's only used for user information, I think we can print what we know. Maybe something like:
Has AccessToken: True/False
AccessTokenExpiry: Date/Unknown (Token Explicitly Specified)

Reply:
Tried to fix this in 0d70853